src/aero_kernel/src/syscall/ipc.rs

```diff
@@ -0,0 +1,161 @@
+}
+
+fn messagequeue_do_recieve(
+    pidptr: usize,
+    messageptr: usize,
+    messagesiz: usize,
+    msg: Message,
+) -> Result<usize, AeroSyscallError> {
+    let output =
+        validate_slice_mut(messageptr as *mut u8, messagesiz).ok_or(AeroSyscallError::EINVAL)?;
+
+    output
+        .split_at_mut(msg.data.len())
+        .0
+        .copy_from_slice(&msg.data);
+
+    unsafe {
+        (pidptr as *mut usize).write(msg.from);
+    }
+    Ok(msg.data.len())
+}
+pub fn send(pid: usize, message: usize, messagesiz: usize) -> Result<usize, AeroSyscallError> {
+    let payload =
+        validate_slice_mut(message as *mut u8, messagesiz).ok_or(AeroSyscallError::EINVAL)?;
+
+    let mut messagequeue = MESSAGES.lock();
+    match messagequeue.get_mut(&pid) {
+        Some(mq) => {
+            mq.queue.push_back(Message {
+                from: get_scheduler().current_task().pid().as_usize(),
+                data: payload.to_vec(),
+            });
+        }
+        None => {
+            messagequeue.insert(
+                pid,
+                MessageQueue {
+                    queue: [Message {
+                        from: get_scheduler().current_task().pid().as_usize(),
+                        data: payload.to_vec(),
+                    }]
+                    .into(),
+                },
+            );
+        }
+    }
+    // let bqueueitem = bqueue.get_mut(&pid).unwrap();
+    // bqueueitem.replace(Message {
+    //     from: get_scheduler().current_task().pid().as_usize(),
+    //     data: payload.to_vec(),
+    // });
+    BLOCK_QUEUE.notify_complete();
+    Ok(0)
+}
+pub fn recv(
+    pidptr: usize,
+    messageptr: usize,
+    messagemax: usize,
+    block: usize,
+) -> Result<usize, AeroSyscallError> {
+    let pid = get_scheduler().current_task().pid().as_usize();
+
+    if block == 0 {
+        // nonblocking read
+        let mut msgqueue = MESSAGES.lock();
+        match msgqueue.get_mut(&pid) {
+            Some(m) => {
+                let item = m
+                    .queue
+                    .pop_front()
+                    .expect("empty message queues should always be deleted!");
+                if item.data.len() > messagemax {
+                    m.queue.push_front(item);
+                    return Err(AeroSyscallError::E2BIG);
+                }
+                if m.queue.len() == 0 {
+                    msgqueue
+                        .remove(&pid)
+                        .expect("safety violation: modification of a value behind a mutex!");
+                }
+                return messagequeue_do_recieve(pidptr, messageptr, messagemax, item);
+            }
+            None => {
+                // nope
+                return Err(AeroSyscallError::EAGAIN);
+            }
+        }
+    }
+
+    let mut queue_map = BLOCK_QUEUE
+        .block_on(&MESSAGES, |msg| {
+            let mp = msg.get_mut(&pid);
+            match mp {
+                Some(mq) => match mq.queue.front() {
```

Comment:
You can replace the `match x { Some(_) => true, None => false, }` with `x.is_some()`

Reply:
okay